### PR TITLE
#3: Extracted cluster configuration and used it in Bucket access.

### LINKED ISSUE #3: Extract random passwords
## Situation

During container creation, random passwords are generated for BucketFS bucket access. While this improves security, it makes using the container harder in integration test scenarios, where the scripts need to know those passwords in order to read from and write to buckets.

The Exasol testcontainer should allow extracting the random passwords.

## Acceptance Criteria

1. Read and write passwords for the default bucket
2. Read and write passwords for any other bucket

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #3.

* Added parser for `EXAConf`
* Abstracted `ClusterConfiguration`
* Added support for bucket access

### CODE REVIEW

File: src/main/java/com/exasol/bucketfs/Bucket.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
+package com.exasol.bucketfs;
+
+import java.io.IOException;
+import java.net.*;
+import java.net.http.*;
+import java.net.http.HttpRequest.BodyPublisher;
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/Bucket.java]: It only imports all types inside the given package (see https://docs.oracle.com/javase/tutorial/java/package/usepkgs.html).

File: src/main/java/com/exasol/bucketfs/Bucket.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+     * This method is intended for writing small objects in BucketFS dynamically like for example configuration files.
+     * For large payload use {@link Bucket#uploadFile(Path, String)} instead.
+     * </p>
+     *
+     * @param pathInBucket path inside the bucket
+     * @param localPath    path of the file to be uploaded
+     * @throws InterruptedException  if the upload is interrupted
+     * @throws BucketAccessException if the file cannot be uploaded to the given URI
+     */
+
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/Bucket.java]: Fixed.

File: src/main/java/com/exasol/bucketfs/Bucket.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    public void uploadStringContent(final String content, final String pathInBucket)
+            throws IOException, InterruptedException, BucketAccessException {
+        final String excerpt = (content.length() > 20) ? content.substring(0, 20) + "..." : content;
+        final URI uri = createWriteURI(pathInBucket);
+        LOGGER.info("Uploading text \"{}\" to \"{}\"", excerpt, uri);
+        try {
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/Bucket.java]: There are subtle differences in all four occurances that make extraction of a common core hard to the point where you lose readability

File: src/main/java/com/exasol/bucketfs/BucketFactory.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    private final ClusterConfiguration clusterConfiguration;
+    private final Map<Integer, Integer> portMappings;
+
+    /**
+     * Create a new instance of a BucketManager.
+     *
+     * @param ipAddress            IP address of the the BucketFS service
+     * @param clusterConfiguration configuration of the Exasol Cluster
+     * @param portMappings         mapping of container internal to exposed port numbers
+     */
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/BucketFactory.java]: Constructor. Intentional.

File: src/main/java/com/exasol/config/BucketConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+ * Configuration parameters of a bucket in BucketFS.
+ */
+public class BucketConfiguration {
+    private final String name;
+    private final boolean publiclyReadable;
+    private final String readPassword;
+    private final String writePassword;
+    private final BucketFsServiceConfiguration serviceConfiguration;
+
+    public BucketConfiguration(final Builder builder) {
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/config/BucketConfiguration.java]: Method should be private. Fixed.

File: src/main/java/com/exasol/config/ClusterConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    private static final String BUCKET_KEY_PREFIX = "Bucket" + KEY_SEPARATOR;
+    private static final String SECTION_SEPARATOR = "/";
+    private static final String DEFAULT_BUCKET_SECTION = "BucketFS:bfsdefault/Bucket:default";
+    private final Map<String, String> parameters;
+
+    /**
+     * Create a new instance of an Exasol {@link ClusterConfiguration}
+     *
+     * @param parameters configuration options as map of keys and values
+     */
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/config/ClusterConfiguration.java]: Constructor. Intentional.

File: src/main/java/com/exasol/config/BucketConfiguration.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/config/BucketConfiguration.java]: Why is it still public?

[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/config/BucketConfiguration.java]: Forgot to push. :-)

File: src/test/java/com/exasol/bucketfs/BucketAccessExceptionTest.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.junit.jupiter.api.Test;
+
+public class BucketAccessExceptionTest {
+    private static final URI EXPECTED_URI = URI.create("http://localhost:2580/default");
+    private static final String EXPECTED_MESSAGE = "the message";
+    private static final Throwable EXPECTED_CAUSE = new IllegalArgumentException();
+
+    @Test
+    public void testCreateFromOtherException() throws Exception {
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/bucketfs/BucketAccessExceptionTest.java]: Is it a Junit 4 test? Junit 5 methods are not public.

File: README.md [hunk 8]
```diff
@@ -43,6 +43,13 @@ This software is intended for use in automated integration tests of Java softwar
 
 * [User Guide](doc/user-guide/user-guide.md)
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply README.md]: Fixed.

File: src/test/java/com/exasol/bucketfs/BucketAccessExceptionTest.java (hunk 7 above)
Comment (redcatbear) [reply src/test/java/com/exasol/bucketfs/BucketAccessExceptionTest.java]: Fixed.

File: doc/design.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+The `ClusterConfiguration` is a set of objects representing the setup of an Exasol cluster.
+
+Needs: impl
+
+## `ConfigurationParser`
+
+The `ConfigurationParser` parses Exasol's cluster configuration format.
+
+
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply doc/design.md]: Fixed.

**Pull request merged**